feat: introducing configurable options

ernolf opened the pull request:
* Solves: #26
* Related: #1324
* Related PR: #1546
* Background: https://help.nextcloud.com/t/lenght-of-totp-key/169226

This is a better, safer approach to introducing configurable options as initially suggested in #1546

This pull request has changed significantly from the initial approach. Please go to https://github.com/nextcloud/twofactor_totp/pull/1549#issuecomment-2278845895 for the description of the current, latest state.
--

**This approach has been significantly improved and abandoned as it was here:**
To ensure that users are not locked out by admin settings, in this approach the values ​​for Token Length and Hash Algorithm can be set individually by each user:

![image](https://github.com/user-attachments/assets/6c492b95-ef1c-4e2b-8fef-1bb1264951e1)

New secrets are always created with the default values ​​because these are supported in every TOTP app. If you use a high-quality TOTP app like Aegis, you can then adjust the values afterwards. This adjustment must be applied identical both here in the server and in the TOTP app.
The administrator still has the option to set the length of the secrte via occ though, as originally suggested in #1546 and described in the README.md:

![image](https://github.com/user-attachments/assets/2fdad670-f9b0-4bbc-acd3-79b0aa4c4565)

---
Finally, the favicon URL is integrated into the setup URL (QR code) so that apps like FreeOTP that support it can use it as an icon for the account:
* Solves: #1407

ernolf

ernolf committed:
feat: introducing configurable options

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: code style fixes applied by php-cs-fixer

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: code style fixes applied by npm run lint -- --fix

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix(tests): adjust SettingsControllerTest setup to match constructor signature

- Added missing IURLGenerator mock to SettingsControllerTest
- Replaced Defaults instance with a mock to match constructor arguments

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix(tests): add missing import for IURLGenerator in SettingsControllerTest

- Added missing use statement for IURLGenerator
- Ensured all required interfaces are properly mocked

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: adjust createSecret method call to match interface definition

- Corrected method call to pass only the IUser argument as required by ITotp::createSecret

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
chore: update generated files after build

- Updated files generated by npm run build to reflect recent changes

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: code style fixes applied by php-cs-fixer

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: add missing use statement and update to use correct constants

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: define getSettings method in ITotp interface

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: INT must be INTEGER

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: use global RuntimeException in Totp service

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: fix and update SettingsController tests for redirect response

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: Fix issues in SettingsControllerTest

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test(tests): fix and run PersonalTotpSettings.spec.js

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
feat: add "Advanced Settings" button

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf commented:
![image](https://github.com/user-attachments/assets/f246c77d-2830-4e8c-8763-3c7c681ba859)

![image](https://github.com/user-attachments/assets/b0876e1b-0dca-4cca-90d9-3b784e610612)

![image](https://github.com/user-attachments/assets/c8bcdca4-87fb-4bea-b031-96a464978205)

![image](https://github.com/user-attachments/assets/3627e40b-d6a1-4049-b6bc-3e3b9b0ff349)

ernolf committed:
fix: correct attribute order in PersonalTotpSettings.vue

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: add missing trailing comma in PersonalTotpSettings.spec.js

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
build: update built JS files after recent changes

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf commented:
It has been questioned whether giving users the option to set the token length and hash algorithm might unnecessarily complicate things for them. It was considered whether these values should be predefined by the admin, like the secret length. In that case, the URL transmitted with the QR code would need to include these values. While the URL can indeed include the token length and hash algorithm along with the secret and issuer, very few TOTP apps actually take these parameters into account. For reference, see the description of the [Key-Uri-Format in the Google Authenticator wiki](https://github.com/google/google-authenticator/wiki/Key-Uri-Format).

In my opinion, it is more important that TOTP works out-of-the-box with ALL TOTP apps. Therefore, it has been designed so that each new secret is ALWAYS set up with the default values, ensuring immediate compatibility with ALL TOTP apps. Users of advanced TOTP apps, such as Aegis, which allow customization of token length and hash algorithm settings, can now take advantage of these options individually.

It is crucial that the values set for token length and/or hash algorithm in the UI are exactly mirrored in the TOTP app for the account. Besides the potential to increase security through a more advanced hash algorithm or a longer token length, the token can also be shortened to as few as 4 digits if desired. Technically, it is even possible to shorten the token length to 1, though this would be too easily compromised. However, a token length of 4 or 5 is still secure, as an attacker would not expect varying token lengths.

The warning message is intentionally very generic and can/should certainly be improved, just like the final buttons. This was primarily about the functionality. The graphics and localization now need to be adjusted to fit the overall look of the Security Page, where, in my opinion, there is still much to improve since everyone seems to be doing their own thing.

ernolf

ernolf committed:
fix: change minlength from 6 to 4 in challenge

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ChristophWurst commented:
I have discussed the feature with the Nextcloud GmbH design lead and product management.

Here are some key aspects to consider
1. The system has to prevent breaking of existing setups. That was the case with the first version. It seems impossible right now :heavy_check_mark: 
2. The app has to be compatible with all popular TOTP apps, like Google Authenticator, by default. Seems this is the case right now :heavy_check_mark: 
3. Configuration for the secret/length and the algorithm must not be done by the user. This has to be an admin setting. :x: 
  4. There has to be a button for advanced settings. In the advanced settings, there is a warning that changing the defaults will limit the number of compatible mobile apps. Possible suggestion: add a list of configuration combinations that are known to work with a specified list of mobile apps.

> It has been questioned whether giving users the option to set the token length and hash algorithm might unnecessarily complicate things for them. It was considered whether these values should be predefined by the admin, like the secret length. In that case, the URL transmitted with the QR code would need to include these values. While the URL can indeed include the token length and hash algorithm along with the secret and issuer, very few TOTP apps actually take these parameters into account. For reference, see the description of the [Key-Uri-Format in the Google Authenticator wiki](https://github.com/google/google-authenticator/wiki/Key-Uri-Format).

I'd say include anything in the QR code string that an app could pick up, as long as it's ignored by the others and doesn't cause compatibility issues.

jancborchardt reviewed (commented):
First off thanks for your contribution @ernolf!

I just have a question or small thing to add to Christoph’s review: Is it possible to use our Vue components for the elements – e.g. the button looks like an old style.
Although I assume that the app in general needs to be ported over so it might be a separate issue @ChristophWurst?

ernolf commented:
> 3\. Configuration for the secret/length and the algorithm must not be done by the user. This has to be an admin setting. ❌

There are a total of three values ​​that this PR is about:

* Secret length
can be changed by the admin and only affects newly created secrets. Secrets that have already been created are not affected.
What still needs to be tested is whether there are any restrictions on this length on the part of Totp client apps.
If the admin does not specify otherwise, the length of 32 characters recommended by the RFC applies.

* OTP tokenlength
Some apps also allow an OTP tokenlength of 8 digits but most apps simply ignore this value when it is transmitted in the URL (QR code).
If this value were to be set by the admin, only very few TOTP apps would remain that would still work.
This PR gives the user the option to set the OTP length in a range from 4 to 10. It must be made clear to the user that this change must also be made in his Totp app. If his Totp app does not allow this, he should leave the values ​​untouched.

* Hash algorithm
This is set to sha1 by default. The simple, most common apps do not provide any way to change this, either with a setting option in the corresponding account in the app or by means of URL (QR code) transfer.
So if the algorithm to be used is set to a different one by the admin, there are only a few Totp apps (I personally only know Aegis) that can handle this and still work.
This PR gives the user the option to choose between sha1, sha256 and sha512 as the hash algorithm. It must be made clear to the user that this change must also be made in his Totp app. If his Totp app does not allow this, he should leave the value ​​untouched.
This option has been requested by users.
However, as soon as one of these last two is changed, there are only a few Totp apps left (I personally only know Aegis) where you can adjust these values. Other apps can not handle those and will stop working.

**Conclusion:**
If this decision is made by the admin in form of a changed default (like with secret-length), the requirement of point 2 (compatibility with all popular TOTP apps) will be broken.
So if the configuration for the OTP-tokenlength and the algorithm must be admin settings, it has to be done for each user separately, which in my opinion would make the feature practicaly useless.
Especially since it is the user who sets up his TOTP app and can also deactivate it again. When deactivating, all settings (secret, hash algorithm and OTP tokenlength) are deleted from the database. When reactivating, the default values ​​start again. So if an admin has entered different values ​​for OTP token length and hash algorithm for a user, then these values ​​are inseparably linked to the secret due to practical constraints. A separate database table that stores such values ​​for each user is possible, but would still break point 2 - not for all users, but for the individual. If, for example, he changes his TOTP app because he is switching from Android to iOS, and wants to set up his new TOTP app, he would most likely run into problems.

> I'd say include anything in the QR code string that an app could pick up, as long as it's ignored by the others and doesn't cause compatibility issues.

Unfortunately, this ignores reality and the way how the TOTP functionality works. If an OTP token length or a hash algorithm that differs from the default is transmitted in the QR code, then it cannot be determined whether these values ​​were adopted by the user's app or ignored. If they are adopted, then Totp works, if they are ignored, then Totp does not work. So unfortunately we have to limit the url (QR-Code) to the lowest common denominator. In other words, exactly as it is currently in use.

> 4\. There has to be a button for advanced settings. In the advanced settings, there is a warning that changing the defaults will limit the number of compatible mobile apps. Possible suggestion: add a list of configuration combinations that are known to work with a specified list of mobile apps.

I think that's exactly what we should focus on here. I already mentioned that I only wrote something generic. Maybe a small project wiki page could even be created where supported apps can be listed.

ernolf

ernolf commented:
> Is it possible to use our Vue components for the elements

Yes @jancborchardt, of course!

This is my first project where I'm going down to the UI level and I have no prior knowledge in these areas. I've had to teach myself practically every step up to this point, but with the Vue components I simply lack the practice and the knowledge of where exactly I need to look.
But I'm looking forward, to learning that too :wink:

jancborchardt commented:
@ernolf cool, no problem – then it's best to do it in a separate follow-up so it's not a blocker for this pull request. :)

ernolf commented:
> 4\. There has to be a button for advanced settings. In the advanced settings, there is a warning that changing the defaults will limit the number of compatible mobile apps. Possible suggestion: add a list of configuration combinations that are known to work with a specified list of mobile apps.

What about this:

![image](https://github.com/user-attachments/assets/a1e65a26-a124-478f-ba40-373d9e4358cf)

@ChristophWurst?

ernolf committed:
refactor: redesign message in advanced setting UI

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
feat: provide favicon as image for FreeOTP

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
style: remove trailing blank line to pass linter

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: add unit test for getFaviconUrl method

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: remove failing test for private method

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: more compact warning in advanced setting UI

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor(totp-settings): fetch advanced settings only when expanded

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
feat(configurable options): Redesign and expansion of configurable options

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

* hashAlgorithm is now 'algorithm'
* tokenLength is now 'digits'
* new configurable option 'period' (lifetime in seconds of the OTP)
* admin can now specify defaults for 'digits' in addition to 'algorithm' and length of the secret
* users can now change settings during setup and enter a custom secret, which will be included in the QR code

ernolf committed:
test: adapted testfiles for configurable options

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
docs: Documentation of configurable options in README.md

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
style: apply code style fixes from Composer

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: remove accidentally included code block

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: move private constants to public interface

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: remove redundant getDefaultPeriod() method

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: adapt test to use static method directly in testCreateSecret

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: add mock for OC.Notification in PersonalTotpSettings tests

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: adjust test for static method invocation

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
style: apply automatic lint fixes

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: change getAlgorithmById from static to instance to make it testable

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: fix QR URL encoding in testCreateSecret

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: comment out sensitive 'secret' parameter in logger

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
feat(ui): improve user feedback and interaction

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

- enhance warning, error, and info messages for clarity
- update mouse behavior for select fields to remove focus on mouseleave
- adjust layout to place select fields vertically instead of horizontally
- apply theme CSS colors for error, warning, and info messages

ernolf committed:
fix: revert PersonalTotpSettings.spec.js to original state from master branch

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: remove redundant Advanced Settings from PersonalTotpSettings.vue

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: update custom event name to kebab-case for Vue.js compliance

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix(tests): remove unused QR import in SetupConfirmation.spec.js

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: use pure vue test suite instead of chai properties

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
style: apply automatic lint fixes

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: remove console logs to comply with linter rules

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
build: update compiled javacript files

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf commented:
README.md:
![image](https://github.com/user-attachments/assets/70161fe2-d350-4e98-a6ca-792286be8b19)

---
no TOTP set up:
![image](https://github.com/user-attachments/assets/c97b66fa-9d3d-4328-91b3-c1a9a7c178cd)

---
Click Enable:
![image](https://github.com/user-attachments/assets/70b40178-8b25-4265-bc14-072af5a067ad)

---
Click "Advanced Settings":
![image](https://github.com/user-attachments/assets/c2927af9-6234-4a04-a3f1-0f92182db8e5)

---
Enter invalid secret:
![image](https://github.com/user-attachments/assets/16e9695a-00b5-48ca-bb0c-6981f55a8668)

The colors (here yelow, blue and red) are defined in the css of your theme:
```
	.warning-message {
		color: var(--color-warning);
	}

	.instruction-message {
		color: var(--color-info);
	}

	.error-message {
		color: var(--color-error);
	}
```

---
If any changes made, "Recreate QR-Code with custom settings" becomes clickable:
![image](https://github.com/user-attachments/assets/67185e51-933b-461b-a2de-7fb260233770)

---
Once clicked, a new QR Code is created and can be scanned:
![image](https://github.com/user-attachments/assets/23832b16-8c4e-4c1b-9d59-36a341d596bd)

---
After verification with an OTP:
![image](https://github.com/user-attachments/assets/48415077-6907-40a7-be47-a657071df33f)

---
The "Advanced Settings" button for subsequent changes is removed now, since it became redundant as it is much easier and safer to apply custom settings via the setup where it can be tested immediately with the OTP.

---
If the QR-code is scanned with Free-OTP or Free-OTP+, the actual favicon wil be used as icon for the account.

---
Since this PR makes it possible to enter a custom secret, #1439 could be closed too.

TODO:
* migration needs to be tested.
* l10n

ernolf committed:
fix(migration): rename 'period' column to 'seconds' to avoid MariaDB syntax conflict

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

mniehren commented:
Would there also be the possibility to set the secret and the other new options over the occ command on console ?

ernolf commented:
> @mniehren
Would there also be the possibility to set the secret and the other new options over the occ command on consol

I've already thought about it, but I wanted to do it in a PR that follows this one. It would build on the functions created here.
I also know of use cases where the admin creates the secrets and should be able to prevent the user from deleting or regenerating them. So I wanted to think about it in detail again so that it doesn't end up being a premature birth.

mniehren commented:
@ernolf 
Great to hear, that's also my use case. The administrator should create the secrets and the user should not be able to modify or delete them, nor should he be able to turn off TOTP. Also the possibility to force TOTP for a particular user, but not for all users, would be great.

ChristophWurst commented:
Just tested again. Impressive work! I'm afraid this goes a bit too far with the complexity. There are advanced settings *for the user*. This was defined as not acceptable

> 3. Configuration for the secret/length and the algorithm must not be done by the user. This has to be an admin setting. ❌

Could you find a compromise where only the admin needs to know about the advanced options?

There are two use cases where I can see stronger options enforced
1) Home user instance for family/friends
2) Big corporation

In both cases it's likely that the person sitting Nextcloud up with TOTP could also tell the users which TOTP app they should use, an fine-tune the settings for the best default that work with the app.

ernolf committed:
chore: merge master into ernolf/configurable_options

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
chore: rebuild assets after merging master

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
feat: add cancel button to abort setup procedure from personal settings

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
feat: integrate advanced settings option in login setup

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: restructure code to satisfy linting rules

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
fix: prevent applying invalid custom secret

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: restructure code to satisfy linting rules

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
test: simplify and improve validation test for custom secret

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
refactor: restructure code to satisfy linting rules

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

ernolf committed:
chore: rebuild assets

Signed-off-by: ernolf <raphael.gradenwitz@googlemail.com>

jancborchardt reviewed (changes requested):
So this is getting very complex now, way too complex even for people who know what they are doing. I also looked again into the original issue #26 to check what the requirements and issues are.

Some points:
- Even if it is behind a feature flag, I would like for the interface to be nice and easy. Otherwise it’s basically just going to be a feature that very few people use, but we need to maintain since it’s in the main app.
- If we want to have it as a "we are so secure" point (which is nice), it shouldn’t be behind a feature flag either.
- Ideal would be to just have 2 settings:
  - "Default (SHA-1)" setting for compatibility with all OTP apps
  - And then the "Enhanced security" setting which uses a better algo with some preset lengths which is also compatible with at least some apps (like Aegis, etc, list of which has to be mentioned next to the setting).
- And it should only be an admin setting as mentioned

ernolf commented:
Hi @jancborchardt, thank you for your reply

> - Ideal would be to just have 2 settings:
>   - "Default (SHA-1)" setting for compatibility with all OTP apps
>   - And then the "Enhanced security" setting which uses a better algo with some preset lengths which is also compatible with at 

As far as I know, all TOTP apps meanwhile support all three hash algorithms supported by *`twofactor_totp`*. Therefore, this is a setting that the admin can set to a default in my solution. If the admin does not set anything, SHA1 applies. The latter is important to ensure that secrets that have already been set up in the past continue to work, because when the database is migrated, these fields for existing secrets are pre-filled with the default, the previous hash algorithm SHA1.

A second default value that the admin can set is the length of the secret. If the admin does not set anything, the standard length of 32 characters applies, which corresponds exactly to the depth of 160 bits recommended by the RFC. The RFC does not allow secrets below 128 bits. For a long time, however, only a 16-character secret was used, which corresponds to a bit depth of 80. So for a long time *`twofactor_totp`* did not even meet the RFC requirements. This is why I have now set the lower limit that the admin can set for the secret to 26 characters, which corresponds to a depth of 130 bits. And I also capped it at 128 characters (640 bits).
Secrets in this range are supported by all apps I know of and therefore this default can be defined by the admin as well.

Then there are two more values:

1. The length of the one-time password

    As far as I know, this can vary between 6 and 8 in every app. In a few apps, however, it can also be shortened to 1 and extended to 10 characters.

    However, since only an OTP token length of 6, 7 or 8 (as far as I know) is supported by all apps, I also made this a default value that can be set by the admin within that range (6-8).

2. The validity period of the one-time password

    This is 30 seconds by default. There are many apps that support a longer validity period, but not all apps support it and there are some apps that support a shorter period. Among those that only support the 30 seconds period is Google Authenticator, which is used by a lot of users. And to guarantee that the admin can't exclude anyone and every app continues to work out of the box without any prior knowledge, I have kept this default value of 30 seconds coded in.

So much for what the admin can set.

> - And it should only be an admin setting as mentioned

There are apps that support other one-time password token lengths outside of these options and allow other validity periods for the one-time passwords. For example, it is very convenient to work with one-time passwords of 4 digits that are valid for 15 seconds. And so that those who want to use this option also have the opportunity to do so, I have now developed the Advanced Settings, which allow the user to set this on an idiot-proof interface that does not allow the user to set or activate a single thing incorrectly.
These settings cannot be predefined by the admin because they go into areas that many apps do not support and are also based on the personal preferences and security requirements of the individual user.

> I also looked again into the original issue #26 to check what the requirements and issues are.

It's no longer just about #26, because that's already been resolved as described in the first paragraph of this answer. It's also about #1439 and I've also solved #1407. The rest are features that I wished for a long time and used so far by haking the code. I am not the type of user that creates feature requests, I create features if I want them and have thought about it a lot. That is what you get

> - Even if it is behind a feature flag, I would like for the interface to be nice and easy. Otherwise it’s basically just going to be a feature that very few people use, but we need to maintain since it’s in the main app.

Is that a subtle hint to fork my solution?
The interface is intentionally designed in such a way that it is absolutely impossible to do anything wrong. Neither incorrect, too short or too long secrets can be adopted, nor can *`twofactor_totp`* be activated incorrectly in any other way, which guarantees that *`twofactor_totp`* can ALWAYS be activated in just a few steps, but never incorrectly. A large part of the complexity of the code is due to this highest security concept and idiot-proofing. Another is that there are actually two interfaces: once when logging in for the first time, when no second factor has been activated yet, and once from the personal settings. Both interfaces have different CSS, so the first is centered and the second is left-aligned, etc.
It is very likely that areas of the code can be simplified. I taught myself every single step of my solution. I have never done frontend development before. So I would be overjoyed if someone who knows vue and can do this off the cuff could help me simplify the code so that it remains maintainable.

> - If we want to have it as a "we are so secure" point (which is nice), it shouldn’t be behind a feature flag either.

Very good! We agree on that point!

Here I have a view of the setup window that is shown when you first log in, once with the advanced settings closed
![image](https://github.com/user-attachments/assets/2981438b-c1b3-41b6-950e-eeba482af320)

and once with the advanced settings opened
![image](https://github.com/user-attachments/assets/11380ebd-a0f4-460c-9a7b-72df9e87c4aa)

I have shortened the labels to make it look simpler. I have also used the same names that are usually used in the TOTP apps so that a clear assignment is possible, although this is not necessary since everything is transmitted with the QR code.

For an extended explanation of the labels and the associated select fields, I have made a :title that is displayed as a tooltip when you mouse over:

example:
```html
			<label :class="{ centered: center }"
				for="digits"
				:title="t('twofactor_totp', 'OTP token length')"
				@mouseleave="onMouseLeave">
				{{ t('twofactor_totp', 'Digits') }}
			</label>
```

that looks like this:

![image](https://github.com/user-attachments/assets/fbb03b9a-569c-4c0d-a426-231d1c4ae398)

> So this is getting very complex now,

Before you get too fixated on "It's all too complex", I would suggest that I demonstrate the feature to you at the Nextcloud conference so that you can "feel" it.

---
ernolf

ChristophWurst commented:
We have integrated the two fundamental changes in https://github.com/nextcloud/twofactor_totp/pull/1778 and https://github.com/nextcloud/twofactor_totp/pull/1779. 

I would still prefer to hide the options from the user because that's very advanced and technical.

Thanks a lot!
